include/cachemere/cache.hpp

```diff
@@ -12,71 +6,83 @@ template<typename Key,
          class InsertionPolicy,
          template<class, class>
          class EvictionPolicy,
+         template<class, class>
+         class ConstraintPolicy,
          typename MeasureValue,
          typename MeasureKey,
          bool ThreadSafe>
-Cache<Key, Value, InsertionPolicy, EvictionPolicy, MeasureValue, MeasureKey, ThreadSafe>::Cache(size_t maximum_size, uint32_t statistics_window_size)
- : m_current_size{0},
-   m_maximum_size{maximum_size},
-   m_statistics_window_size{statistics_window_size},
-   m_insertion_policy(std::make_unique<InsertionPolicy<Key, Value>>()),
+template<typename... Args>
+Cache<Key, Value, InsertionPolicy, EvictionPolicy, ConstraintPolicy, MeasureValue, MeasureKey, ThreadSafe>::Cache(Args... args)
+ : m_insertion_policy(std::make_unique<InsertionPolicy<Key, Value>>()),
    m_eviction_policy(std::make_unique<EvictionPolicy<Key, Value>>()),
-   m_measure_key{},
-   m_measure_value{},
+   m_constraint_policy(std::make_unique<ConstraintPolicy<Key, Value>>(std::forward<Args>(args)...)),
    m_mutex{},
    m_data{},
-   m_hit_rate_acc(boost::accumulators::tag::rolling_window::window_size = statistics_window_size),
-   m_byte_hit_rate_acc(boost::accumulators::tag::rolling_window::window_size = statistics_window_size)
+   m_hit_rate_acc(boost::accumulators::tag::rolling_window::window_size = m_statistics_window_size),
+   m_byte_hit_rate_acc(boost::accumulators::tag::rolling_window::window_size = m_statistics_window_size)
 {
 }
 
-template<class K, class V, template<class, class> class I, template<class, class> class E, class SV, class SK, bool TS>
-template<typename C>
-Cache<K, V, I, E, SV, SK, TS>::Cache(C& collection, size_t maximum_size, uint32_t statistics_window_size) : Cache(maximum_size, statistics_window_size)
+template<class K, class V, template<class, class> class I, template<class, class> class E, template<class, class> class C, class SV, class SK, bool TS>
+template<typename Coll, typename... Args>
+Cache<K, V, I, E, C, SV, SK, TS>::Cache(Coll& collection, std::tuple<Args...> args)
+ : m_insertion_policy(std::make_unique<I<K, V>>()),
+   m_eviction_policy(std::make_unique<E<K, V>>()),
+   m_constraint_policy(
+       std::move(std::apply([](auto&&... params) { return std::make_unique<C<K, V>>(std::forward<decltype(params)>(params)...); }, std::move(args)))),
```

Comment:
Yup. It's because the other constructor also has variadic arguments, so it causes ambiguity. 

I'm _really_ not a fan of passing this tuple though, so I'm all ears if you have ideas to improve this bit.